# <center> <font face="Times New Roman" size=5 > Pandas 資料處理 </font> </center>

## <font face="Times New Roman" size=4 > 一、先安裝個Pandas套件 </font>

In [3]:
pip install pandas

In [2]:
#安裝好後看一下版本
import pandas as pd
pd.__version__

'2.2.1'

## <font face="Times New Roman" size=4 > 二、建立 DataFrame </font>

*   <font face="Times New Roman" size=4 >建立 DataFrame</font>
*   <font face="Times New Roman" size=4 >讀取 CSV 檔</font>

### <font face="Times New Roman" size=4 > (一)、建立 DataFrame </font>

In [ ]:
#直接建立DataFrame

salary_list=pd.DataFrame({
    "name":["Amy","John","Bob"],
    "salary":[30000,50000,40000]
})

In [6]:
'''
用Dict建立DataFrame
Dict的格式是Key-Value Key=欄位 Value=資料

'''


dic = {
    "col 1": [1, 2, 3], 
    "col 2": [10, 20, 30],
    "col 3": list('xyz'),
    "col 4": ['a', 'b', 'c'],
    "col 5": pd.Series(range(3))
}

df = pd.DataFrame(dic)
df

col 1  col 2 col 3 col 4  col 5
0      1     10     x     a      0
1      2     20     y     b      1
2      3     30     z     c      2

### <font face="Times New Roman" size=4 > (二)、讀取 CSV 檔 </font>

In [4]:
df = pd.read_csv('C:\\Users\\9921916\\Desktop\\test\\aqx_p_07.csv')
df.head()

sitename siteengname areaname county township    siteaddress    twd97lon  \
0       大城     Dacheng    中部空品區    彰化縣      大城鄉   彰化縣大城鄉西厝路98號  120.269642   
1      富貴角  FugueiCape    北部空品區    新北市      石門區            NaN  121.536569   
2       麥寮     Mailiao   雲嘉南空品區    雲林縣      麥寮鄉  雲林縣麥寮鄉中興路115號  120.251825   
3       關山    Guanshan    花東空品區    臺東縣      關山鎮   臺東縣關山鎮自強路66號  121.161933   
4       馬公      Magong       其他    澎湖縣      馬公市  澎湖縣馬公市中正路115號  119.566158   

    twd97lat sitetype  siteid  
0  23.854931      其它站      85  
1  25.296817      背景站      84  
2  23.753506      工業站      83  
3  23.045083      其它站      80  
4  23.569031      一般站      78

In [5]:
# 可以稍微看一下這堆資料沾了記憶體多少空間
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sitename     78 non-null     object 
 1   siteengname  78 non-null     object 
 2   areaname     78 non-null     object 
 3   county       78 non-null     object 
 4   township     78 non-null     object 
 5   siteaddress  77 non-null     object 
 6   twd97lon     78 non-null     float64
 7   twd97lat     78 non-null     float64
 8   sitetype     78 non-null     object 
 9   siteid       78 non-null     int64  
dtypes: float64(2), int64(1), object(7)
memory usage: 6.2+ KB


## <font face="Times New Roman" size=4 > 三、數據清理 </font>

### <font face="Times New Roman" size=4 > (一)空值處理 </font>

https://medium.com/@zector1030/pandas-fillna-%E7%AF%84%E4%BE%8B-5d33819fb7b8

In [30]:
'''
利用 fillna 函式將 DataFrame 裡頭所有不存在的值設為 0
或是可以設定成自己容易辨別的字串:

df.fillna("Unknown")

'''

df['siteaddress'].fillna(value="Unknown", inplace=True)
df.head()


C:\Users\9921916\AppData\Local\Temp\ipykernel_10916\828281736.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['siteaddress'].fillna(value="Unknown", inplace=True)


sitename siteengname areaname county township    siteaddress    twd97lon  \
0       大城     Dacheng    中部空品區    彰化縣      大城鄉   彰化縣大城鄉西厝路98號  120.269642   
1      富貴角  FugueiCape    北部空品區    新北市      石門區        Unknown  121.536569   
2       麥寮     Mailiao   雲嘉南空品區    雲林縣      麥寮鄉  雲林縣麥寮鄉中興路115號  120.251825   
3       關山    Guanshan    花東空品區    臺東縣      關山鎮   臺東縣關山鎮自強路66號  121.161933   
4       馬公      Magong       其他    澎湖縣      馬公市  澎湖縣馬公市中正路115號  119.566158   

    twd97lat sitetype  siteid  
0  23.854931      其它站      85  
1  25.296817      背景站      84  
2  23.753506      工業站      83  
3  23.045083      其它站      80  
4  23.569031      一般站      78

In [ ]:
'''
如果是數值型資料可以用平均值取代 nan (或是其他方法)

'''

df['欄位'].fillna(value=df['欄位'].mean(), inplace=True)
df

### <font face="Times New Roman" size=4 > (二)取得想要的數據 </font>

*   <font face="Times New Roman" size=4 >loc 含頭也含尾</font>
*   <font face="Times New Roman" size=4 >iloc 含頭不含尾</font>

> column name 取資料的時候用 loc，用 index 取資料的時候用 iloc

https://ticyyang.medium.com/python%E8%B3%87%E6%96%99%E5%88%86%E6%9E%90-pandas%E7%9A%84-loc%E8%88%87iloc%E4%B9%8B%E7%95%B0%E5%90%8C-bf7d62541676

In [35]:
# 例如我要取第 2~5 的 row：
df_loc = df.loc[2:5]
print(df_loc)

# df_iloc = df.iloc[2:6]
# print(df_iloc)


  sitename siteengname areaname county township    siteaddress    twd97lon  \
2       麥寮     Mailiao   雲嘉南空品區    雲林縣      麥寮鄉  雲林縣麥寮鄉中興路115號  120.251825   
3       關山    Guanshan    花東空品區    臺東縣      關山鎮   臺東縣關山鎮自強路66號  121.161933   
4       馬公      Magong       其他    澎湖縣      馬公市  澎湖縣馬公市中正路115號  119.566158   
5       金門      Kinmen       其他    金門縣      金城鎮   金門縣金城鎮民權路32號  118.312256   

    twd97lat sitetype  siteid  
2  23.753506      工業站      83  
3  23.045083      其它站      80  
4  23.569031      一般站      78  
5  24.432133      一般站      77  
  sitename siteengname areaname county township    siteaddress    twd97lon  \
2       麥寮     Mailiao   雲嘉南空品區    雲林縣      麥寮鄉  雲林縣麥寮鄉中興路115號  120.251825   
3       關山    Guanshan    花東空品區    臺東縣      關山鎮   臺東縣關山鎮自強路66號  121.161933   
4       馬公      Magong       其他    澎湖縣      馬公市  澎湖縣馬公市中正路115號  119.566158   
5       金門      Kinmen       其他    金門縣      金城鎮   金門縣金城鎮民權路32號  118.312256   

    twd97lat sitetype  siteid  
2  23.753506      工業站     

In [41]:
# 以column name取資料

# column name沒加上中括號，資料格式為Series
df.loc[:, 'township']    

# column name加上中括號，資料格式為DataFrame
df.loc[:, ['township']]    

#取多個column
df.loc[:, ['areaname', 'township', 'siteid']]

areaname township  siteid
0     中部空品區      大城鄉      85
1     北部空品區      石門區      84
2    雲嘉南空品區      麥寮鄉      83
3     花東空品區      關山鎮      80
4        其他      馬公市      78
..      ...      ...     ...
73    北部空品區      土城區       5
74    北部空品區      新店區       4
75    北部空品區      萬里區       3
76    北部空品區      汐止區       2
77    北部空品區      信義區       1

[78 rows x 3 columns]

In [49]:
# 以index取資料

# index沒加上中括號，資料格式為Series
df.iloc[:, 2]     

# index加上中括號，資料格式為DataFrame
df.iloc[:, [2]]     

# 取多個 column
# 1.連續
df.iloc[:, 2:4] 

# 2.不連續
df.iloc[:, [2, 6, 4]]  

areaname    twd97lon township
0     中部空品區  120.269642      大城鄉
1     北部空品區  121.536569      石門區
2    雲嘉南空品區  120.251825      麥寮鄉
3     花東空品區  121.161933      關山鎮
4        其他  119.566158      馬公市
..      ...         ...      ...
73    北部空品區  121.451861      土城區
74    北部空品區  121.537778      新店區
75    北部空品區  121.689881      萬里區
76    北部空品區  121.640810      汐止區
77    北部空品區  121.760056      信義區

[78 rows x 3 columns]

### <font face="Times New Roman" size=4 > (三)刪除不要的欄位 </font>

In [6]:
columns = ['siteengname', 'twd97lat']
df.drop(columns, axis=1)

sitename areaname county township                siteaddress    twd97lon  \
0        大城    中部空品區    彰化縣      大城鄉               彰化縣大城鄉西厝路98號  120.269642   
1       富貴角    北部空品區    新北市      石門區                        NaN  121.536569   
2        麥寮   雲嘉南空品區    雲林縣      麥寮鄉              雲林縣麥寮鄉中興路115號  120.251825   
3        關山    花東空品區    臺東縣      關山鎮               臺東縣關山鎮自強路66號  121.161933   
4        馬公       其他    澎湖縣      馬公市              澎湖縣馬公市中正路115號  119.566158   
..      ...      ...    ...      ...                        ...         ...   
73       土城    北部空品區    新北市      土城區            新北市土城區學府路一段241號  121.451861   
74       新店    北部空品區    新北市      新店區              新北市新店區民族路108號  121.537778   
75       萬里    北部空品區    新北市      萬里區              新北市萬里區瑪鋉路221號  121.689881   
76       汐止    北部空品區    新北市      汐止區  221新北市汐止區北市汐止區樟樹一路137巷26號  121.640810   
77       基隆    北部空品區    基隆市      信義區                 基隆市東信路324號  121.760056   

   sitetype  siteid  
0       其它站      85  
1       背景站      84  
2       工業站      83  
3       其它站      80  
4       一般站      78  
..      ...     ...  
73      一般站       5  
74      一般站       4  
75      一般站       3  
76      一般站       2  
77      一般站       1  

[78 rows x 8 columns]

### <font face="Times New Roman" size=4 > (四)切割成多個欄位 </font>

In [19]:
# 字串

df = pd.DataFrame({
    "name": ["大雄", "胖虎"], 
    "feature": ["膽小, 翻花繩", "粗魯, 演唱會"]
})
df

name  feature
0   大雄  膽小, 翻花繩
1   胖虎  粗魯, 演唱會

In [16]:
df[['性格', '特技']] = df.feature.str.split(',', expand=True)
df

name  feature  性格    特技
0   大雄  膽小, 翻花繩  膽小   翻花繩
1   胖虎  粗魯, 演唱會  粗魯   演唱會

In [35]:
#list
df = pd.DataFrame({
    "name": ["大雄", "胖虎"], 
    "feature": [["膽小", "翻花繩"], ["粗魯", "演唱會"]]
})

df

name    feature
0   大雄  [膽小, 翻花繩]
1   胖虎  [粗魯, 演唱會]

In [38]:
cols = ['性格', '特技']
df_feature = pd.DataFrame(df.feature.tolist(), columns=cols)

df_name = df.iloc[:,[0]]
# df = pd.concat([df_name, df_feature], join='outer', ignore_index=True)
df = pd.concat([df_name, df_feature], axis=1)

df

name  性格   特技
0   大雄  膽小  翻花繩
1   胖虎  粗魯  演唱會

In [28]:
#也可以用apply來切割

df.feature.apply(pd.Series)

0    1
0  膽小  翻花繩
1  粗魯  演唱會

### <font face="Times New Roman" size=4 > (五)新增資料 </font>
https://blog.csdn.net/sinat_29957455/article/details/84961936

>List可用 

In [43]:
data1 = {
  "age": [16, 14, 10],
  "qualified": [True, True, True]
}
df1 = pd.DataFrame(data1)

data2 = {
  "age": [55, 40],
  "qualified": [True, False]
}
df2 = pd.DataFrame(data2)

newdf = df1.append(df2)

print(newdf)

AttributeError: 'DataFrame' object has no attribute 'append'

### <font face="Times New Roman" size=4 > Cleaning Data Practice </font>

In [46]:
df_book = pd.read_csv('C:\\Users\\9921916\\Desktop\\test\\Book1.csv')
df_book

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60    2020/12/26    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

### <font face="Times New Roman" size=4 > 一、刪除空值，並將Calories的空值填上平均數 </font>

In [47]:
df_book['Calories'].fillna(value=df_book['Calories'].mean(), inplace=True)

new_df = df_book.dropna()   #結果會存在新的表單
# df_book.dropna()            #會更改df_book
print(new_df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'  

C:\Users\9921916\AppData\Local\Temp\ipykernel_14780\3083933879.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_book['Calories'].fillna(value=df_book['Calories'].mean(), inplace=True)


### <font face="Times New Roman" size=4 > 二、更正時間格式錯誤 </font>

In [ ]:
df_book['Date'] = pd.to_datetime(df_book['Date'])

print(df_book.to_string())

### <font face="Times New Roman" size=4 > 三、更正下方資料錯誤 </font>

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

> Duration is 45 in row 7

In [ ]:
df_book.loc[7, 'Duration'] = 45

### <font face="Times New Roman" size=4 > 四、刪除重複的資料 </font>

df.duplicated()

# <center> <font face="Times New Roman" size=5 > END </font> </center>